packages/react/src/component.tsx

```diff
@@ -264,45 +272,62 @@ export const UploadDropzone = <
       </div>
     </div>
   );
-};
+}
 
-export const Uploader = <TRouter extends void | FileRouter = void>(props: {
-  endpoint: EndpointHelper<TRouter>;
-  onClientUploadComplete?: () => void;
-  url?: string;
-}) => {
+export function Uploader<TRouter extends FileRouter>(
+  props: FileRouter extends TRouter
+    ? ErrorMessage<"You forgot to pass the generic">
+    : UploadthingComponentProps<TRouter>,
+) {
   return (
     <>
       <div className="flex flex-col items-center justify-center gap-4">
         <span className="text-center text-4xl font-bold">
           {`Upload a file using a button:`}
         </span>
+        {/* @ts-expect-error - this is validated above */}
         <UploadButton<TRouter> {...props} />
       </div>
       <div className="flex flex-col items-center justify-center gap-4">
         <span className="text-center text-4xl font-bold">
           {`...or using a dropzone:`}
         </span>
+        {/* @ts-expect-error - this is validated above */}
         <UploadDropzone<TRouter> {...props} />
       </div>
     </>
   );
-};
+}
+
+function Spinner() {
+  return (
+    <svg
+      className="ut-animate-spin ut-h-5 ut-w-5 ut-text-white"
+      xmlns="http://www.w3.org/2000/svg"
+      fill="none"
+      viewBox="0 0 576 512"
+    >
+      <path
+        fill="currentColor"
+        d="M256 32C256 14.33 270.3 0 288 0C429.4 0 544 114.6 544 256C544 302.6 531.5 346.4 509.7 384C500.9 399.3 481.3 404.6 465.1 395.7C450.7 386.9 445.5 367.3 454.3 351.1C470.6 323.8 480 291 480 255.1C480 149.1 394 63.1 288 63.1C270.3 63.1 256 49.67 256 31.1V32z"
+      />
+    </svg>
+  );
+}
 
 export function generateComponents<TRouter extends FileRouter>() {
   return {
-    UploadButton: (
-      props: React.ComponentProps<typeof UploadButton<TRouter>>,
-    ) => {
-      return <UploadButton {...props} />;
+    UploadButton: (props: UploadthingComponentProps<TRouter>) => {
+      // @ts-expect-error - this is validated above
```

Comment:
Definetely don't like that there is 5 `@ts-expect-error` in this file but it does still work as it should:

![CleanShot 2023-06-15 at 22 43 35](https://github.com/pingdotgg/uploadthing/assets/51714798/b8e8d097-9db5-4b08-88c7-3523a4542a3a)

Would love to fix this though